Flannel integration enhancements and bug fixes

aanm opened the pull request:
Read per commit basis.

A follow up PR will be made scripts to cleanup BPF programs upon uninstall.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/6726)
<!-- Reviewable:end -->

aanm committed:
plugins/cilium-cni: rely on flannel response

With CNI chaining, the response from the previous CNI plugin, flannel,
is sent to the previous CNI plugin, Cilium, we can rely on the
flannel response to create a Cilium endpoint instead of entering the
network namespace of the container.

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
daemon: append bpf programs to existing flannel running containers

As Cilium has the ability to install BPF programs in already running
interfaces we can enable this option by default environments where
Cilium is deployed on top of Kubernetes. This requires Cilium DaemonSet
to run with `hostPID: true` so that cilium-agent process can access
already running network namespaces.

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
pkg/datapath: ignore iptables MASQUERADE rules in flannel mode

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
plugins/cilium-cni: change CNI channing name to cbr0

As flannel runs with the name "cbr0" by default in its cni
configuration, in order for Cilium to be able to run on top of flannel
with existing containers, the flannel-cilium CNI configuration file
needs to have the same name. By having the same name it prevents the
flannel IPAM assigning existing addresses as they are stored on each
node under the directory `/var/lib/cni/networks/cbr0/`.

Signed-off-by: André Martins <andre@cilium.io>

ianvernon commented on daemon/daemon_main.go:
```diff
@@ -1102,7 +1102,7 @@ func runDaemon() {
 			log.WithError(err).WithField("device", option.Config.FlannelMasterDevice).Fatal("Unable to set internal IPv4")
 		}
-		// d.attachExistingInfraContainers()
+		d.attachExistingInfraContainers()
```
What will happen when this code runs and the Cilium DaemonSet is not running with `hostPID: true`?

aanm replied:
Currently yes, I'll think on something to avoid this.

aanm committed:
daemon: stop waiting for flannel device to be ready if process is terminated

When running in flannel mode, cilium-agent would ignore the sigterms as
it would be infinitely waiting for the flannel interface to be ready.

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
pkg/client: wait until the client has connectivity with daemon

Fixes: c6518190de30 ("client: Add NewDefaultClientWithTimeout function")
Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
addons/flannel: add more flannel documentation

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
plugins/cilium-cni: uninstall cilium-flannel CNI configuration if user opt-in

This avoids containers to accidentally be created and starting being
managed only by Flannel if Cilium is in the process of being upgraded
where the CNI configuration is always removed during the upgrade.

Signed-off-by: André Martins <andre@cilium.io>

ianvernon commented on examples/kubernetes/1.10/cilium-crio-ds.yaml:
```diff
@@ -121,6 +121,12 @@ spec:
               key: ct-global-max-entries-other
               name: cilium-config
               optional: true
+        - name: CILIUM_PREALLOCATE_BPF_MAPS
```
I'm surprised this was missing from the generated file before.

ianvernon commented on pkg/client/client.go:
```diff
@@ -86,14 +86,18 @@ func NewDefaultClientWithTimeout(timeout time.Duration) (*Client, error) {
 		c, err = NewDefaultClient()
 		if err == nil {
-			break
+			_, err := c.Daemon.GetConfig(nil)
```
Not-blocking: would be helpful to add a comment indicating that we make a request against the daemon here purely to see if it has connectivity (i.e., calling `GetConfig` vs. some other API call doesn't matter).

ianvernon commented on examples/kubernetes/addons/flannel/README.md:
```diff
@@ -4,8 +4,12 @@ Cilium integration with Flannel (EXPERIMENTAL)
+install Flannel again.
+
+This Cilium integration with Flannel was performed with Flannel 0.10.0. If you
+find any issues with previous Flannel versions please feel free to reach us out
```
Nit: reach out to us to help you
